### PR TITLE
Added first unit tests and bug fixes

### PR DESCRIPTION
**Pull request by apfong**
Fixed mismatched curly brackets and nested blocks.

### CODE REVIEW

**Comment by kfrankc** [comment]
`d3405da` should be split into two commits. One for fixing the parser for unpaired and nested bracket, and one for the unit tests.

File: config_parser.cc [hunk 1]
```diff
@@ -128,7 +128,7 @@ NginxConfigParser::TokenType NginxConfigParser::ParseToken(std::istream* input,
         *value += c;
         continue;
       case TOKEN_STATE_TOKEN_TYPE_NORMAL:
-        if (c == ' ' || c == '\t' || c == '\n' || c == '\t' ||
+        if (c == ' ' || c == '\t' || c == '\n' ||
```
Comment (kfrankc) [inline config_parser.cc:131]: This is very nice, I did not even notice the extra condition in the `if` statements

File: config_parser_test.cc [hunk 2]
```diff
@@ -9,3 +12,184 @@ TEST(NginxConfigParserTest, SimpleConfig) {
...
+                            # hello world\n\
+                            listen 80;\n\
+                          }"));
+  EXPECT_TRUE(ParseString("server { # simple reverse-proxy\n\
+                             listen 80;\n\
+                           }"));
+}
+
+TEST_F(NginxStringConfigTest, FullConfig) {
+  EXPECT_TRUE(ParseString("user  www www;\n\
```
Comment (kfrankc) [inline config_parser_test.cc]: These long strings could be put into files, making it easier to view + to test. Just a future practice

File: config_parser_test.cc [hunk 3]
```diff
@@ -9,3 +12,184 @@ TEST(NginxConfigParserTest, SimpleConfig) {
...
+TEST_F(NginxStringConfigTest, OnlyCurlyConfig) {
+  EXPECT_FALSE(ParseString("{}"));
+}
+
+TEST_F(NginxStringConfigTest, NoBlockNameConfig) {
+  EXPECT_FALSE(ParseString("{ listen 80; }"));
+}
+
+TEST_F(NginxStringConfigTest, CommentConfig) {
+  EXPECT_TRUE(ParseString("# hello world\n server { listen 80; }"));
```
Comment (kfrankc) [inline config_parser_test.cc:88]: I suggest adding comments on failure: ex. `<< "Error: failed to detect comment";` on these tests for more clarity when they are fired.

**Review by kfrankc — CHANGES_REQUESTED** [review]
Added two comment on the test fixtures

File: .gitignore [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+!*.*
+
+# Unignore all dirs
+!*/
+
+# Unignore config files
+!*_config
+
+### Above combination will ignore all files without extension ###
+# From http://stackoverflow.com/questions/5711120/gitignore-without-binary-files
```
Comment (KyCodeHuynh) [inline .gitignore:14]: A very useful gitignore!

File: config_parser_test.cc [hunk 5]
```diff
@@ -9,3 +12,184 @@ TEST(NginxConfigParserTest, SimpleConfig) {
 
   EXPECT_TRUE(success);
 }
+
+TEST(NginxConfigParserTest, VhostsConfig) {
+  NginxConfigParser parser;
+  NginxConfig out_config;
+
+  bool success = parser.Parse("vhosts_config", &out_config);
```
Comment (KyCodeHuynh) [inline config_parser_test.cc]: Where is this `vhosts_config` file? I know your gitignore above ignores files without extensions, but this test doesn't appear to be repeatable without the config files used by each test?

File: config_parser_test.cc [hunk 6]
```diff
@@ -9,3 +12,184 @@ TEST(NginxConfigParserTest, SimpleConfig) {
...
+
+  EXPECT_TRUE(success);
+}
+
+TEST(NginxConfigParserTest, ListenDefaultConfig) {
+  NginxConfigParser parser;
+  NginxConfig out_config;
+
+  bool success = parser.Parse("listen_default_config", &out_config);
```
Comment (KyCodeHuynh) [inline config_parser_test.cc]: [Similar to above](https://github.com/apfong/nginx-configparser/pull/1/files#r96323085)

**Review by KyCodeHuynh — CHANGES_REQUESTED** [review]
Looks great overall, but it looks like some test input is missing.

File: config_parser_test.cc (hunk 5 above)
Comment (apfong) [reply config_parser_test.cc]: Fixed, see [commit](https://github.com/apfong/nginx-configparser/pull/1/commits/af237643145f3b6c5b8b4f7b146b119cde13021a)

File: config_parser_test.cc (hunk 6 above)
Comment (apfong) [reply config_parser_test.cc]: [Same commit as previous comment](https://github.com/apfong/nginx-configparser/pull/1#discussion_r96325335)

File: config_parser_test.cc (hunk 2 above)
Comment (apfong) [reply config_parser_test.cc]: [Got it](https://github.com/apfong/nginx-configparser/pull/1/commits/435bd2d8a54da461ee5c0ff4656ec46b3a755ea2)

File: config_parser_test.cc (hunk 3 above)
Comment (apfong) [reply config_parser_test.cc:88]: [Error messages are in](https://github.com/apfong/nginx-configparser/pull/1/commits/435bd2d8a54da461ee5c0ff4656ec46b3a755ea2)

File: config_parser_test.cc (hunk 5 above)
Comment (KyCodeHuynh) [reply config_parser_test.cc]: Awesome, thank you!

**Review by KyCodeHuynh — APPROVED** [review]
Great work! LGTM now. 👍

**Review by kfrankc — APPROVED** [review]
Looks good!

**Pull request merged**